django/db/backends/base/features.py

```diff
@@ -38,13 +38,19 @@ class BaseDatabaseFeatures:
     # integer primary keys.
     related_fields_match_type = False
     allow_sliced_subqueries_with_in = True
+    has_select_for_share = False
+    has_select_for_share_nowait = False
+    has_select_for_share_skip_locked = False
+    has_select_for_share_of = False
+    has_select_for_key_share = False
     has_select_for_update = False
     has_select_for_update_nowait = False
     has_select_for_update_skip_locked = False
     has_select_for_update_of = False
     has_select_for_no_key_update = False
-    # Does the database's SELECT FOR UPDATE OF syntax require a column rather
-    # than a table?
+    # Does the database's SELECT FOR SHARE/UPDATE OF syntax require a column
+    # rather than a table?
+    select_for_share_of_column = False
```

Comment:
There is a lot of noise in this patch related with renaming tests and feature flags. I'd:
- rename the current feature flags in a separate PR (with a short release note) and used them for both methods:
  - `has_select_for_update_nowait` -> `has_select_for_with_nowait`,
  - `has_select_for_update_skip_locked` -> `has_select_for_with_skip_locked`
  - `has_select_for_update_of` -> `has_select_for_with_of`
- move `tests/select_for_update/tests.py` to `tests/select_for/test_for_update.py` in a separate PR and add `tests/select_for/test_for_share.py` in this PR. It may be easier to maintain similar tests in both files and we will be able to avoid tricks like:
  ```python
  if not getattr(connection.features, flag):
      continue  # XXX: Cannot skip subtests, see bpo-35327.
  ```

Reply:
Thanks for the guidance, that's exactly what I was looking for at this stage to avoid a lot of further effort in the wrong direction... 🙂

Will rebase to 5.0 and split into separate commits in this pull request first before splitting into separate pull requests.

> - move `tests/select_for_update/tests.py` to `tests/select_for/test_for_update.py` in a separate PR and add `tests/select_for/test_for_share.py` in this PR... 

I started this nearly three years ago and it had some early rough decisions. I had figured that a bunch of tests could share code, especially around the flag handling. This might still be possible with a mixin as there is quite some duplication. I had started down the sub test route and then realised that the output was broken when trying to skip a sub test. Separate modules looks like the way to go 👍🏻 

> - rename the current feature flags in a separate PR (with a short release note) and used them for both methods:

Thanks, but I'm not sure this makes sense? Some backends _may_ not have parity of flag implementation for both locking modes... I'm thinking about third-party backends in particular that we don't necessarily have a good understanding of.

Here is a table of the support for the main major databases:

|                               | MariaDB       | MySQL     | Oracle    | PostgreSQL    | SQLite    | MSSQL     |
| ----------------------------- | ------------- | --------- | --------- | ------------- | --------- | --------- |
| `FOR SHARE`                   | 🗸[^1]         | 🗸[^1]     | ×         | 🗸 ≥8.1        | ×         | 🗸[^2].    |
| • Supported with `LIMIT`?     | -             | -         | ?         | -             | -         | -         |
| • After `FROM` clause?        | -             | -         | -         | -             | -         | 🗸         |
| `FOR SHARE OF`                | ×             | 🗸 ≥8.0.1  | ?         | 🗸 ≥8.1        | ×         | ×         |
| • Requires column, not table? | -             | -         | ?         | -             | -         | -         |
| `FOR SHARE WAIT [n]`          | 🗸 ≥10.3       | ×         | ?         | ×             | ×         | ×         |
| `FOR SHARE NOWAIT`            | 🗸[^6] ≥10.3   | 🗸 ≥8.0.1  | ?         | 🗸 ≥8.1        | ×         | 🗸[^4]     |
| `FOR SHARE SKIP LOCKED`       | 🗸[^7] ≥10.6   | 🗸 ≥8.0.1  | ?         | 🗸 ≥9.5        | ×         | 🗸[^5]     |
| `FOR KEY SHARE`               | ×             | ×         | ?         | 🗸 ≥9.3        | ×         | ×         |
| `FOR UPDATE`                  | 🗸             | 🗸         | 🗸         | 🗸             | ×         | 🗸[^3]     |
| • Supported with `LIMIT`?     | -             | -         | ×         | -             | -         | -         |
| • After `FROM` clause?        | -             | -         | -         | -             | -         | 🗸         |
| `FOR UPDATE OF`               | ×             | 🗸 ≥8.0.1  | 🗸         | 🗸             | ×         | ×         |
| • Requires column, not table? | -             | -         | 🗸         | -             | -         | -         |
| `FOR UPDATE WAIT [n]`         | 🗸 ≥10.3       | ×         | 🗸         | ×             | ×         | ×         |
| `FOR UPDATE NOWAIT`           | 🗸[^6] ≥10.3   | 🗸 ≥8.0.1  | 🗸         | 🗸 ≥8.1        | ×         | 🗸[^4]     |
| `FOR UPDATE SKIP LOCKED`      | 🗸[^7] ≥10.6   | 🗸 ≥8.0.1  | 🗸         | 🗸 ≥9.5        | ×         | 🗸[^5]     |
| `FOR NO KEY UPDATE`           | ×             | ×         | ×         | 🗸 ≥9.3        | ×         | ×         |

On this basis, the support for various options _does_ seem to match up between the share and update locks. The only question would be whether this would hold true if Oracle were to ever implement row-level share locks... 🤔  

_Would you still like me to go ahead with renaming the flags?_

---

[^1]: Alternative `SELECT … LOCK IN SHARE MODE` syntax required for MariaDB and MySQL < 8.0.1.
[^2]: Alternative `SELECT … FROM … WITH (ROWLOCK, HOLDLOCK) …` syntax required.
[^3]: Alternative `SELECT … FROM … WITH (ROWLOCK, UPDLOCK) …` syntax required.
[^4]: Append `NOWAIT` as a table hint in `WITH (…)`.
[^5]: Append `READPAST` as a table hint in `WITH (…)`.
[^6]: `NOWAIT` is equivalent to `WAIT 0` on MariaDB.
[^7]: Only supported when using InnoDB on MariaDB.

Reply:
> _Would you still like me to go ahead with renaming the flags?_

OK, let's be generous here :wink: We can keep extra flags.

Reply:
Cool. I guess it also avoids extra work for third-party backends having to "if django version is greater than ... else ..." a load of stuff.